server/src/service/OfferService.cc

```diff
@@ -0,0 +1,135 @@
+#include "OfferService.hh"
+
+namespace server::service {
+
+    oatpp::Object<server::dto::OfferDto>
+    OfferService::createOne(
+            oatpp::Object<server::dto::OfferDto> const &dto
+    ) {
+        try {
+            this->getOne(dto->id); // Will throw 404 if not found
+            OATPP_ASSERT_HTTP(false, Status::CODE_409, "Offer already exists")
+        }
+        catch (oatpp::web::protocol::http::HttpError &e) {
+            if (e.getInfo().status != Status::CODE_404) { throw; }
+        }
+
+        auto queryResult = database_->createOffer(dto);
+
+        OATPP_ASSERT_HTTP(queryResult->isSuccess(), Status::CODE_500, queryResult->getErrorMessage())
+        OATPP_ASSERT_HTTP(queryResult->hasMoreToFetch(), Status::CODE_500, "No rows returned!")
+
+        auto fetchResult = queryResult->fetch<oatpp::Vector<oatpp::Object<dto::OfferDto>>>();
+
+        OATPP_ASSERT_HTTP(fetchResult->size() == 1, Status::CODE_500, "Unexpected number of rows returned!")
+
+        return fetchResult[0];
+    }
+
+    oatpp::Object<server::dto::OfferDto>
+    OfferService::getOne(
+            oatpp::UInt64 const &id
+    ) {
+        auto queryResult = database_->getOffer(id);
+
+        OATPP_ASSERT_HTTP(queryResult->isSuccess(), Status::CODE_500, queryResult->getErrorMessage())
+        OATPP_ASSERT_HTTP(queryResult->hasMoreToFetch(), Status::CODE_404, "Offer not found")
+
+        auto fetchResult = queryResult->fetch<oatpp::Vector<oatpp::Object<dto::OfferDto>>>();
+
+        OATPP_ASSERT_HTTP(fetchResult->size() == 1, Status::CODE_500, "Unexpected number of rows returned!")
+
+        return fetchResult[0];
+    }
+
+    oatpp::Object<dto::OfferPageDto>
+    OfferService::search(
+            oatpp::String const &query,
+            oatpp::UInt64 const &limit,
+            oatpp::UInt64 const &offset
+    ) {
+        std::string const queryTableFts = " FROM offer_fts";
+        std::string const queryFilters = query->empty() ? std::string{} : " WHERE offer_fts MATCH :query";
```

Comment:
We can use `std::string_view` or `char const*` to avoid allocating heap memory.

Reply:
Can you show me what do you mean?

Reply:
```
std::string_view queryTableFts = " FROM offer_fts";
std::string_view queryFilters = query->empty() ? "" : " WHERE offer_fts MATCH :query";
```

Reply:
fixed in 37a52bdc4aaaef38964d8bb8634d82a981c667a8